2023-03-24_Transactional-Outbox-and-Dapr-PubSub/transactional-outbox-and-dapr-pub-sub.md

```diff
@@ -0,0 +1,150 @@
+
+```csharp
+async (CompleteOrderDto dto, ICompleteOrderUnitOfWork unitOfWork) =>
+{
+    var order = await unitOfWork.GetOrderAsync(dto.OrderId);
+    if (order is null)
+        return Results.NotFound();
+    
+    order.State = OrderState.Complete;
+    var outboxItem = OrderCompletedEvent.CreateFromOrder(order); 
+    unitOfWork.Add(outboxItem);
+    await unitOfWork.SaveChangesAsync();
+    
+    return Results.Ok(order);
+}
+```
+
+In the code above, we only inject a unit-of-work and update our order as well as insert an outbox item to the database. We then create a new component that reads outbox items from the database and publishes them to the message broker. A simplified version might look like this:
+
+```csharp
+public sealed class OutboxProcessor
+{
+    public OutboxProcessor(IOutboxUnitOfWork unitOfWork,
+                           IEventFactory eventFactory,
+                           IEventPublisher eventPublisher,
+                           ILogger<OutboxProcessor> logger)
+    {
+        UnitOfWork = unitOfWork;
+        EventFactory = eventFactory;
+        EventPublisher = eventPublisher;
+        Logger = logger;
+    }
+
+    private IOutboxUnitOfWork UnitOfWork { get; }
+    private IEventFactory EventFactory { get; }
+    private IEventPublisher EventPublisher { get; }
+    private ILogger<OutboxProcessor> { get; }
+    
+    // Other members that trigger the actual functionality are left out for brevity's sake
+    
+    private async Task ProcessAsync()
+    {
+        var random = new Random();
+        while (true)
+        {
+            try
+            {
+                var outboxItems = await unitOfWork.GetNextOutboxItemsAsync();
+                if (outboxItems.Count == 0)
+                    return;
+                
+                foreach (var outboxItem in outboxItems)
+                {
+                    var @event = EventFactory.CreateEventFromOutboxItem(outboxItem);
+                    await EventPublisher.PublishEventAsync(@event);
+                    unitOfWork.Remove(outboxItem);
+                    await unitOfWork.SaveChangesAsync();
+                }
+            }
+            catch (Exception exception)
+            {
+                Logger.LogError(exception, "An error occurred while processing the outbox");
+                var waitTimeInMilliseconds = random.Next(1000, 3000);
+                await Task.Delay(waitTimeInMilliseconds);
+            }
+        }
+    }
+}
+```
+
+In the piece of code above, we can see an outbox processor whose `ProcessAsync` loads a batch of outbox items from the database and then transforms and sends each item to the message broker. If an error occurs, the processor restarts from the beginning of the loop until all outbox items are sent. It can be triggered periodically (e.g. every few seconds) or other components can trigger it, e.g. when they have created and saved an outbox item.
+
```

Comment:
A proper vehicle for running the `OutboxProcessor` could be a Kubernetes `DaemonSet` which will process all outbox items created on the given worker-node.